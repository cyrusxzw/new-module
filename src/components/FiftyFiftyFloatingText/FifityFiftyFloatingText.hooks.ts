import { useEffect, useState } from 'react';
import { isViewport } from '~/utils/viewport';
import { ActiveViewTypes } from './FiftyFiftyFloatingText.types';
import { useWindowHasResized } from '~/customHooks';

const useActiveView = (): { activeView: ActiveViewTypes } => {
  const [activeView, setActiveView] = useState<ActiveViewTypes>('none');

  useWindowHasResized();

  const isMobile = isViewport('xs to sm only');
  const isTabletAndDesktop = isViewport('md');

  useEffect(() => {
    if (isMobile) {
      setActiveView('mobile');
    } else {
      setActiveView('tabletanddesktop');
    }
  }, [isMobile, isTabletAndDesktop]);

  return { activeView };
};

export { useActiveView };
