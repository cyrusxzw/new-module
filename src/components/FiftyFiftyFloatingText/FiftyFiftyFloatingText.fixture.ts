import type { FiftyFiftyFloatingTextProps } from './FiftyFiftyFloatingText.types';

const FiftyFiftyFloatingTextFixture: FiftyFiftyFloatingTextProps = {
  backgroundImages: [
    {
      image:
      {
        altText: 'alt image',
        sizes: {
          large: './assets/images/MediaWithContent/fifty-fifty-img1.png',
          medium: './assets/images/MediaWithContent/fifty-fifty-img1.png',
          small: './assets/images/MediaWithContent/fifty-fifty-img1.png',
        },
      }
    },
    {
      image:
      {
        altText: 'alt image',
        sizes: {
          large: './assets/images/MediaWithContent/fifty-fifty-img2.png',
          medium: './assets/images/MediaWithContent/fifty-fifty-img2.png',
          small: './assets/images/MediaWithContent/fifty-fifty-img2.png',
        },
      }
    },
    {
      image:
      {
        altText: 'alt image',
        sizes: {
          large: './assets/images/MediaWithContent/fifty-fifty-img3.png',
          medium: './assets/images/MediaWithContent/fifty-fifty-img3.png',
          small: './assets/images/MediaWithContent/fifty-fifty-img3.png',
        },
      }
    },
  ],
  id: 'fifty-fifty-floating-text',
  textBlocks: [
    {
      text:
        '1. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.',
    },
    {
      text:
        '2. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.',
    },
    {
      text:
        '3. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.',
    },
  ],
  theme: 'light',
};

export { FiftyFiftyFloatingTextFixture };
