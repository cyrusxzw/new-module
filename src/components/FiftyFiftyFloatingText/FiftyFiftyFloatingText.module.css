@import 'styles/index';

.base {
  position: relative;
  display: grid;
  margin-top: 0;
  background-color: #d9d8cf;
  grid-template-columns: 1fr 1fr;

  @media (--viewport-xs-sm-only) {
    display: block;
  }
}

.scrollLockMedia {
  position: sticky;
  top: 0;
  left: 0;
  height: 100vh;

  @media (--viewport-md-only) {
    margin-right: 20px;
  }
}

.scrollingContent {
  position: relative;
}

@media all and (-ms-high-contrast: none) {
  .scrollingContent {
    margin-top: 0;
  }
  .scrollLockMedia {
    position: absolute;
    height: 100%;
    clip: rect(auto, auto, auto, auto);
  }
  .backgroundElement {
    position: fixed;
    top: 0;
    left: 0;
  }
}
