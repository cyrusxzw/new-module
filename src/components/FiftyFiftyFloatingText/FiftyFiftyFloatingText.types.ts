import type { Themes } from '~/types';
import type { ImageSectionProps } from './components/ImageSection/ImageSection.types';
import type { TextSectionProps } from './components/TextSection/TextSection.types';

type FiftyFiftyFloatingTextProps = {
  backgroundImages?: {
    image?: ImageSectionProps['image'];
  }[];
  id?: string;
  percentageOfViewport?: number;
  textBlocks?: {
    text?: TextSectionProps['text'];
  }[];
  theme?: Themes;
};

type ActiveViewTypes = 'none' | 'mobile' | 'tabletanddesktop';

export type { FiftyFiftyFloatingTextProps, ActiveViewTypes };
