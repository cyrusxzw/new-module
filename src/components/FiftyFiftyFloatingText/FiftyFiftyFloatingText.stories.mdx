import { Fragment } from 'react';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import {
  StorybookGridOverlay,
  StorybookNavHeader,
} from '~.storybook/components';
import { FiftyFiftyFloatingText } from './FiftyFiftyFloatingText';
import { FiftyFiftyFloatingTextFixture } from './FiftyFiftyFloatingText.fixture';

<Meta
  title="Components/FiftyFiftyFloatingText"
  component={FiftyFiftyFloatingText}
/>

### Fifty Fifty Floating Text

<Canvas>
  <Story name="Fifty Fifty Floating Text">
    {(() => {
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
        navHeader: {
          isActive: knobs.boolean(
            'Show navigation header',
            false,
            'Implementation',
          ),
          theme: knobs.select(
            'Navigation header theme',
            ['dark', 'light'],
            'dark',
            'Implementation',
          ),
        },
      };
      return (
        <>
          <StorybookNavHeader
            theme={implementation.navHeader.theme}
            isActive={implementation.navHeader.isActive}
          />
          <FiftyFiftyFloatingText
            backgroundImages={FiftyFiftyFloatingTextFixture.backgroundImages}
            id={FiftyFiftyFloatingTextFixture.id}
            textBlocks={FiftyFiftyFloatingTextFixture.textBlocks}
            theme={knobs.select(
              'theme',
              ['dark', 'light'],
              'dark',
              'Presentation',
            )}
          />
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

# API

<ArgsTable of={FiftyFiftyFloatingText} />
