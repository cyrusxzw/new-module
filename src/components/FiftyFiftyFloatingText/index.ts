import { FiftyFiftyFloatingText } from './FiftyFiftyFloatingText';

export { FiftyFiftyFloatingText };
