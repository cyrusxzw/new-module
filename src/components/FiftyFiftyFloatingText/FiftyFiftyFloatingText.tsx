import React, { useEffect, forwardRef, useRef, useState } from 'react';
import cx from 'classnames';
import { isInBrowser } from '~/utils/environment';
import { ImageSection } from './components/ImageSection';
import { TextSection } from './components/TextSection';
import styles from './FiftyFiftyFloatingText.module.css';
import type { FiftyFiftyFloatingTextProps } from './FiftyFiftyFloatingText.types';

const FiftyFiftyFloatingText = forwardRef<any, FiftyFiftyFloatingTextProps>(
  (
    {
      backgroundImages = [],
      id,
      textBlocks = [],
      theme = 'light',
      percentageOfViewport = 0.4,
    },
    ref,
  ) => {
    const revealRefs = useRef([]);
    const addToRefs = (el) => {
      if (el && !revealRefs.current.includes(el)) {
        revealRefs.current.push(el);
      }
    };

    const [textHeight, setTextHeight] = useState();

    const handleScroll = () => {
      if (isInBrowser() && textHeight) {
        const images = Array.from(
          document.getElementsByClassName(
            'fiftyScrollImage',
          ) as HTMLCollectionOf<HTMLElement>,
        );

        let opacity = null;
        revealRefs.current.forEach((ele, index) => {
          const eleToTop = ele.getBoundingClientRect().top; // the distance of each element to top

          // offset = 0 (element become in viewport) offset > textHeight (element hidden on top) offset < 0 (element hidden on bottom)
          const offset = textHeight - eleToTop;
          if (offset > 0 && offset < textHeight) {
            // text reach percentageOfViewport (default is 40%) of viewport, image opacity: 100%
            const opacityTemp = offset / (percentageOfViewport * textHeight);
            opacity = opacityTemp >= 1 ? 1 : opacityTemp; // control max opacity is 100%
            images[index].style.opacity = opacity;
          }
        });
      }
    };

    useEffect(() => {
      window.addEventListener('scroll', handleScroll);
      return function cleanup() {
        window.removeEventListener('scroll', handleScroll);
      };
    });

    const classSet = cx(styles.base, styles[theme]);
    const classImageSection = cx(styles.backgroundElement, 'fiftyScrollImage');

    return (
      <section className={classSet} id={id} ref={ref}>
        <div className={cx(styles.scrollLockMedia)}>
          {backgroundImages.map(({ image }, index) => (
            <ImageSection
              className={classImageSection}
              image={image}
              key={index}
            />
          ))}
        </div>
        <div className={styles.scrollingContent}>
          {textBlocks.map(({ text }, index) => (
            <TextSection
              key={index}
              ref={addToRefs}
              setTextHeight={setTextHeight} // get the height of text block wrapper from TextSection
              text={text}
              theme={theme}
            />
          ))}
        </div>
      </section>
    );
  },
);

export { FiftyFiftyFloatingText };
