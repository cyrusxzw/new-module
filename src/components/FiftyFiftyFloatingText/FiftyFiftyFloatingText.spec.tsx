import React from 'react';
import { render } from '@testing-library/react';
import { FiftyFiftyFloatingText } from './FiftyFiftyFloatingText';
import { FiftyFiftyFloatingTextFixture } from './FiftyFiftyFloatingText.fixture';

describe('<FullWidthHeroScroll />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <FiftyFiftyFloatingText
        textBlocks={FiftyFiftyFloatingTextFixture.textBlocks}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
