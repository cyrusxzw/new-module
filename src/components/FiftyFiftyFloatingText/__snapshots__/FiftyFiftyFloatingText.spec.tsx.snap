// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FullWidthHeroScroll /> should render base component correctly 1`] = `
<div>
  <section
    class="base light"
  >
    <div
      class="scrollLockMedia"
    />
    <div
      class="scrollingContent"
    >
      <div
        class="base light"
      >
        <div>
          1. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.
        </div>
      </div>
      <div
        class="base light"
      >
        <div>
          2. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.
        </div>
      </div>
      <div
        class="base light"
      >
        <div>
          3. Aliquam turpis libero, posuere sed nisl in, porta facilisis lorem. Phasellus eu turpis in quam efficitur hendrerit.
        </div>
      </div>
    </div>
  </section>
</div>
`;
